ostd/src/mm/page_table/node/mod.rs

```diff
@@ -152,135 +138,56 @@ impl<E: PageTableEntryTrait, C: PagingConstsTrait> RawPageTableNode<E, C> {
     pub(super) unsafe fn first_activate(&self) {
         use crate::{arch::mm::activate_page_table, mm::CachePolicy};
 
-        self.inc_ref_count();
-
-        activate_page_table(self.raw, CachePolicy::Writeback);
+        activate_page_table(self.clone().into_raw(), CachePolicy::Writeback);
     }
+}
 
-    fn inc_ref_count(&self) {
-        // SAFETY: We have a reference count to the page and can safely increase the reference
-        // count by one more.
-        unsafe {
-            inc_frame_ref_count(self.paddr());
-        }
-    }
+/// A reference to a page table node.
+pub(super) type PageTableNodeRef<'a, E, C> = FrameRef<'a, PageTablePageMeta<E, C>>;
 
-    /// Restores the handle from the physical address and level.
+impl<'a, E: PageTableEntryTrait, C: PagingConstsTrait> PageTableNodeRef<'a, E, C> {
+    /// Creates a new [`PageTableGuard`] without checking if the page table lock is held.
     ///
     /// # Safety
     ///
-    /// The caller must ensure that the physical address is valid and points to
-    /// a forgotten page table node. A forgotten page table node can only be
-    /// restored once. The level must match the level of the page table node.
-    pub(super) unsafe fn from_raw_parts(paddr: Paddr, level: PagingLevel) -> Self {
-        Self {
-            raw: paddr,
-            level,
-            _phantom: PhantomData,
-        }
-    }
-}
-
-impl<E: PageTableEntryTrait, C: PagingConstsTrait> From<RawPageTableNode<E, C>>
-    for Frame<PageTablePageMeta<E, C>>
-{
-    fn from(raw: RawPageTableNode<E, C>) -> Self {
-        let raw = ManuallyDrop::new(raw);
-        // SAFETY: The physical address in the raw handle is valid and we are
-        // transferring the ownership to a new handle. No increment of the reference
-        // count is needed.
-        unsafe { Frame::<PageTablePageMeta<E, C>>::from_raw(raw.paddr()) }
-    }
-}
-
-impl<E: PageTableEntryTrait, C: PagingConstsTrait> Drop for RawPageTableNode<E, C> {
-    fn drop(&mut self) {
-        // SAFETY: The physical address in the raw handle is valid. The restored
-        // handle is dropped to decrement the reference count.
-        drop(unsafe { Frame::<PageTablePageMeta<E, C>>::from_raw(self.paddr()) });
+    /// This function must be called if this task logically holds the lock.
+    ///
+    /// Calling this function when a guard is already created is undefined behavior
+    /// unless that guard was already forgotten.
+    pub(super) unsafe fn make_guard_unchecked(self) -> PageTableGuard<'a, E, C> {
```

Comment:
Similarly:
```rust
    pub(super) unsafe fn make_guard_unchecked(self, _guard: &'rcu dyn InAtomicMode) -> PageTableGuard<'rcu, E, C> {
```

The safety requirement claims that the guard exists, but that does not have too much to with the RCU lifetime. But anyway, in our cases the caller should have the guard, so we can just pass the guard instead of adding another safety requirement.